### DISCUSSION TITLE
How to constrain pulse amplitude?

### DISCUSSION THREAD

**Discussion by will-atom** [JuliaQuantumControl/QuantumControl.jl#41] [Q&A] [answered]
Is it possible to constrain a pulse amplitude to be less than a fixed maximum value? For example for an instance of `ShapedAmplitude`? 

Thanks! 
-Will

**Comment by goerz** [comment, marked as answer]
Yes, see https://juliaquantumcontrol.github.io/Krotov.jl/stable/examples/state_to_state_parametrizations/.

For Krotov, such parametrization is the only way to constrain the pulse amplitude.

The GRAPE method also has the option of specifying bounds, see https://juliaquantumcontrol.github.io/GRAPE.jl/stable/api/#QuantumControlBase.optimize-Tuple{Any,%20Val{:GRAPE}}

Be aware though that such bounds can lead to traps in the optimization landscape, so you might have to play around a bit with different guess pulses.

**Reply by HuelsMat** [reply]
Hi,

Couldn't we achieve the same by defining a suitable `update_hook` function that we pass as an argument,

 `optimize(problem; method=Krotov,update_hook)`?
 
Am I misunderstanding `update_hook`?

In my application, I would like to process my pulses after each iteration, e.g. apply a spectral filter and cut amplitudes to allowed values. Following the [documentation](https://juliaquantumcontrol.github.io/QuantumControl.jl/stable/methods/), `update_hook` should do the job.

**Reply by goerz** [reply]
You could, but there's no guarantee that it won't throw you off track in terms of convergence. Of course, that applies to the spectral filtering just as well. So yeah, you can do any kind of filtering of the pulses after each iteration. If it works, it works, but the more you change the pulse compared to the update that Krotov wants, the more likely you are to not converge. For GRAPE there might be additional issues if you use LBFGS as a backend (which is the default): because LBFGS builds up a history of gradient information to estimate the Hessian, if you change the control in inconsistent ways, that might screw up the search direction.

BTW, as a heads-up: I have plans to combine `update_hook` and `info_hook` into a `callback` in the somewhat near future, so you may have to adapt any code using `update_hook` at that time.

**Reply by HuelsMat** [reply]
I see, Thanks!

Would you recommend a different method to avoid high frequencies and large amplitudes?

I'm afraid that a pulse parametrization will restrict my search space too much.
Maybe GRAPE is better suited, as I can set limits for the amplitudes. 
But here the optimized pulses are anything but smooth and I yet don't see how to avoid sharp edges (high frequencies).

**Reply by goerz** [reply]
With Krotov, increase `lambda_a`
